Allow pre-binding of Persistent Volumes to PVClaims

markturansky opened the pull request:
@thockin @saad-ali This small PR is necessary per our discussion about dynamic provisioning, but I think it's a good change overall for PVs.  It would work for any type of PV, regardless of how it was provisioned.

Allow a PV to be intended for 1 specific claim by adding the annotation:  

```
pv.Annotations[createdForKey] = "namespace/claimName"
```

An admin can use this for any PV they are managing.

markturansky commented:
@kubernetes/rh-storage

saad-ali commented on pkg/controller/persistentvolume/types.go:
```diff
@@ -25,6 +25,17 @@ import (
+	// A PVClaim containing this annotation requests a quality of service tier.
+	// The value of this field is arbitrary and known by the user and administrator.
+	// This is an experimental feature and likely to change in the future.
+	qualityOfServiceRequestKey = "experimental/k8s.io/persistentVolumeProvisioning-qos"
```
Since it's not yet used anywhere, let's remove this (it can be re-added along with code using it).

markturansky replied:
Sure thing.  I will re-introduce this with the provisioner.  It will look for this on a PVC and provision a volume for it and add the 'createdForKey' annotation.

saad-ali commented:
LGTM mod nit

markturansky committed:
allow pre-binding of persistent volumes to pvclaims

markturansky commented:
thanks @saad-ali.  I removed the annotation, as requested.

I also changed the value of the 'createdFor' annotation to fit what Tim recently wrote for https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#label-selector-and-annotation-conventions

saad-ali commented on pkg/controller/persistentvolume/types.go:
```diff
@@ -25,6 +25,13 @@ import (
+	// A PV created specifically for one claim must contain this annotation in order to bind to the claim.
+	// The value must be the name of the claim being bound to.
+	// This is an experimental feature and likely to change in the future.
+	createdForKey = "provisioning.experimental.kubernetes.io/createdFor"
```
nit: the name `provisioning` doesn't make it clear that it is PV provisioning (not load balancer provisioning, for example).

markturansky replied:
good point.  i also saw (after submitting the recent name change) that Tim called for dashes instead of camelCasing in the API doc.   

Key value updated to reflect PV specifically and it uses hot looking dashes instead of the old and busted camelCase.

saad-ali commented:
LGTM nit on the name, I'll leave it up to you if you want to change it

markturansky commented:
Different flake in Shippable.  1 of 2 shippable builds passed.

`F0923 13:12:14.155971   16215 integration.go:936] FAILED: (Scheduler probably didn't process deletion of 'phantom.bar') Pod never started running: timed out waiting for the condition`

markturansky commented:
@k8s-bot test this

markturansky commented:
Same Shippable error as I posted previously.

Go 1.4 contains the error (as last time) while Go 1.3 does not.

brendandburns commented on pkg/controller/persistentvolume/types.go:
```diff
@@ -25,6 +25,13 @@ import (
+	// A PV created specifically for one claim must contain this annotation in order to bind to the claim.
+	// The value must be the name of the claim being bound to.
+	// This is an experimental feature and likely to change in the future.
+	createdForKey = "persistent-volume-provisioning.experimental.kubernetes.io/created-for"
```
we should centralize these tag names somewhere.  (pkg/apis/experimental/annotations.go?)

markturansky replied:
I can move them if that's where we'd prefer annotations to live.  Do we care about scoping at all?  These constants package protected and aren't relevant outside of this narrow functionality.

The docs explaining this feature will require the keys, of course.

thockin replied:
that's pretty wordy.  how about volume.experimental.kubernetes.io/provisioned-for

markturansky replied:
This whole PV feature has been wordy ... but I can type "persistent" very
quickly now, so I've got that going for me.

I'll change it to your suggestion :)

On Wed, Sep 23, 2015 at 5:46 PM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/controller/persistentvolume/types.go
> <https://github.com/kubernetes/kubernetes/pull/14249#discussion_r40261952>
> :
>
> > @@ -25,6 +25,13 @@ import (
> >  	"k8s.io/kubernetes/pkg/client/cache"
> >  )
> >
> > +const (
> > +	// A PV created specifically for one claim must contain this annotation in order to bind to the claim.
> > +	// The value must be the name of the claim being bound to.
> > +	// This is an experimental feature and likely to change in the future.
> > +	createdForKey = "persistent-volume-provisioning.experimental.kubernetes.io/created-for"
>
> that's pretty wordy. how about
> volume.experimental.kubernetes.io/provisioned-for
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14249/files#r40261952>.
>

brendandburns commented on pkg/controller/persistentvolume/types.go:
```diff
@@ -93,16 +100,32 @@ func (pvIndex *persistentVolumeOrderedIndex) Find(pv *api.PersistentVolume, matc
+				if createdFor != searchPV.Annotations[createdForKey] {
+					// the volume is pre-bound and does not match the search criteria.
+					continue
+				} else {
```
remove 'else'

brendandburns commented:
LGTM, we're in merge freeze while we try to stabilize Shippable.  This is fine to come in after we get out of that.

markturansky committed:
shortened annotation key name, improved comments

markturansky commented:
'else' removed, annotation key name changed to suggestion, and comments improved.
